hazelcast/src/main/java/com/hazelcast/internal/management/request/ExecuteScriptRequest.java

```diff
@@ -73,47 +61,53 @@ public int getType() {
 
     @Override
     public void writeResponse(ManagementCenterService mcs, JsonObject root) throws Exception {
-        JsonObject jsonResult = new JsonObject();
-        ArrayList results;
-        if (targetAllMembers) {
-            Set<Member> members = mcs.getHazelcastInstance().getCluster().getMembers();
-            ArrayList<Object> list = new ArrayList<Object>(members.size());
-            for (Member member : members) {
-                list.add(mcs.callOnMember(member, new ScriptExecutorOperation(engine, script, bindings)));
-            }
-            results = list;
-        } else {
-            ArrayList<Object> list = new ArrayList<Object>(targets.size());
-            for (String address : targets) {
-                AddressUtil.AddressHolder addressHolder = AddressUtil.getAddressHolder(address);
-                Address targetAddress = new Address(addressHolder.getAddress(), addressHolder.getPort());
-                list.add(mcs.callOnAddress(targetAddress, new ScriptExecutorOperation(engine, script, bindings)));
-            }
-            results = list;
+        Map<Address, Future<Object>> futures = MapUtil.createHashMap(targets.size());
+
+        for (String address : targets) {
+            AddressUtil.AddressHolder addressHolder = AddressUtil.getAddressHolder(address);
+            Address targetAddress = new Address(addressHolder.getAddress(), addressHolder.getPort());
+            futures.put(targetAddress, mcs.callOnAddress(targetAddress, new ScriptExecutorOperation(engine, script)));
         }
 
-        StringBuilder sb = new StringBuilder();
-        for (Object result : results) {
-            if (result instanceof String) {
-                sb.append(result);
-            } else if (result instanceof List) {
-                List list = (List) result;
-                for (Object o : list) {
-                    sb.append(o).append("\n");
-                }
-            } else if (result instanceof Map) {
-                Map map = (Map) result;
-                for (Object o : map.entrySet()) {
-                    Map.Entry entry = (Map.Entry) o;
-                    sb.append(entry.getKey()).append("->").append(entry.getValue()).append("\n");
+        JsonObject responseJson = new JsonObject();
+        StringBuilder scriptResult = new StringBuilder();
+        for (Map.Entry<Address, Future<Object>> entry : futures.entrySet()) {
+            Address address = entry.getKey();
+            Future<Object> future = entry.getValue();
+
+            try {
+                Object result = future.get();
```

Comment:
I went back and forth between creating a custom class for returning the call's success/failure output together with its result/error message (since Java doesn't allow multiple return values from a method) and leaving it to `Future` handling code. In the end, I chose to go with leaving it to `Future` handling code. Just to make it clear, you think the `ManagementCenterService` method should resolve the `Future` and return the status and the message to its callers?

Reply:
Not necessarily - I just mean composing a string and error handling required because of `future.get()` are two things and better done one after another.